Fixed the bug of resuing mesh with different shaders.

rongguodong opened the pull request:
Now, we generate VBO for different shaders and record them in a map.

GearVRf-DCO-1.0-Signed-off-by: Guodong Rong
rongguodong@hotmail.com

JonShemitz commented on GVRf/Framework/jni/objects/mesh.h:
```diff
@@ -30,344 +30,322 @@
+		return iterator != vaoID_map_.end() ? iterator->second : 0;
+	}
+
+	void cleanUp() {
```
I think if you move this right after `~Mesh()`, the diff will be much easier to read.

rongguodong commented:
I have just moved cleanup() to right below ~mesh() following Jon's comment.

I noticed that this function seems to be introduced by the commit of frustum culling. They just simply duplicated all codes in the destructor in a new function cleanup(). I removed the duplication by calling this function in the destructor in this commit, so that it is easier to maintain.

JonShemitz commented:
@chenchao1407 ?

JonShemitz commented on GVRf/Framework/jni/objects/mesh.cpp:
```diff
@@ -166,6 +166,9 @@ void Mesh::generateVAO() {
     }
 
+    GLuint vaoID_ = 0;
+	GLuint triangle_vboID_, vert_vboID_, norm_vboID_, tex_vboID_;
```
I don't know if you intend this indenting to signify anything ... if so, it won't survive a Shift+Ctrl+F

rongguodong replied:
I do not see any indenting on my PC. And it seems Shift+Ctrl+F does not work correctly for CPP files. Maybe my formatter setting has any problems?

JonShemitz replied:
Maybe. I’ve had success formatting .h and .cpp files – it’s the *_jni.cpp files that don’t work well.

From: Guodong Rong [mailto:notifications@github.com]
Sent: Thursday, May 28, 2015 11:29 AM
To: Samsung/GearVRf
Cc: Jon Shemitz
Subject: Re: [GearVRf] Fixed the bug of resuing mesh with different shaders. (#73)


In GVRf/Framework/jni/objects/mesh.cpp<https://github.com/Samsung/GearVRf/pull/73#discussion_r31263207>:

> @@ -166,6 +166,9 @@ void Mesh::generateVAO() {

>          return;

>      }

>

> +    GLuint vaoID_ = 0;

> +     GLuint triangle_vboID_, vert_vboID_, norm_vboID_, tex_vboID_;

I do not see any indenting on my PC. And it seems Shift+Ctrl+F does not work correctly for CPP files. Maybe my formatter setting has any problems?

—
Reply to this email directly or view it on GitHub<https://github.com/Samsung/GearVRf/pull/73/files#r31263207>.

JonShemitz commented on GVRf/Framework/jni/objects/mesh.cpp:
```diff
@@ -247,6 +250,12 @@ void Mesh::generateVAO() {
     }
 
+    vaoID_map_[key] = vaoID_;
+	triangle_vboID_map_[key] = triangle_vboID_;
```
I may be missing something, but `triangle_vboID_`, `vert_vboID_` and `tex_vboID_` never seem to be set in these diffs.

rongguodong replied:
They are set in the original files, not shown in diff.

JonShemitz commented on GVRf/Framework/jni/objects/mesh.h:
```diff
@@ -30,344 +30,322 @@
+	}
+
+	void cleanUp() {
+		std::vector<glm::vec3> vertices;
```
Most of the differences in this file appear to just be whitespace ... which makes it very hard to see what you really changed.

rongguodong commented:
I pushed a new commit after doing Ctrl-Shift-F for .cpp and .h files. It seems the original files have different format with mine. That's the reason of so many "blank" changes. :(

rongguodong commented:
At least, there is no strange indenting in mesh.cpp any more. :)

thomasflynn commented on GVRf/Framework/jni/objects/mesh.h:
```diff
@@ -30,344 +30,322 @@
-        return vaoID_;
-    }
-
-    void cleanUp() {
```
it looks like cleanUp() got deleted.  intentional?

rongguodong replied:
No. It is moved up right below the destructor according to Jon's suggestion.

thomasflynn commented on GVRf/Framework/jni/shaders/material/bounding_box_shader.cpp:
```diff
@@ -28,65 +28,65 @@
-    if (program_ != 0) {
-        recycle();
-    }
+	if (program_ != 0) {
```
not just in this file, but in general, it looks like your tabs are set to 8?

rongguodong commented:
It turns out that I did not set "use space only" for C++ formatter. Now it should be much better.

BTW: I tried a batch formatting for all Java code, and found out 840+ files are changed (most, if not all, changes are in Java Doc comments). So I have to abandoned that change, and only manually do Ctrl-Shift-F for all touched .cpp and .h files in this commit.

chenchao1407 commented on GVRf/Framework/jni/objects/mesh.h:
```diff
@@ -354,8 +329,11 @@ class Mesh: public HybridObject {
+    std::map<Material::ShaderType, GLuint> triangle_vboID_map_;
+    std::map<Material::ShaderType, GLuint> vert_vboID_map_;
+    std::map<Material::ShaderType, GLuint> norm_vboID_map_;
+    std::map<Material::ShaderType, GLuint> tex_vboID_map_;
```
Just my minor concern here. As far as I can see, actually only we only need map for vaoIDs since only vaoID has getter which requires key. For all other vboID, we just need some kind of container to store them and finally call glDeleteBuffer with all of them.  In std, map is actually a tree(which may differentiate from hash_map), which may be a bit more expensive then something like a dequeue or vector.

JonShemitz replied:
> In std, map is actually a tree(which may differentiate from hash_map), which may be a bit more expensive then something like a dequeue or vector.

More memory consumption, yes. Slower, probably not.

chenchao1407 commented:
Tried with sample cubemap and pickandmove, works as expected with no side effect. I think it is good to go.

JonShemitz commented:
>  I think it is good to go.

Thanks, @chenchao1407  - I'll take a closer look than I did yesterday.
